src/morpho-blue/VirtualHealthFactor.sol

```diff
@@ -0,0 +1,104 @@
+import {MorphoLib} from "../../lib/morpho-blue/src/libraries/periphery/MorphoLib.sol";
+import {MathLib} from "../../lib/morpho-blue/src/libraries/MathLib.sol";
+
+import {ORACLE_PRICE_SCALE} from "../../lib/morpho-blue/src/libraries/ConstantsLib.sol";
+
+/// @title Morpho Blue Snippets
+/// @author Morpho Labs
+/// @custom:contact security@morpho.org
+/// @notice The Virtual Health Factor Snippets contract.
+contract VirtualHealthFactorSnippets {
+    using MathLib for uint256;
+    using MorphoLib for IMorpho;
+    using MorphoBalancesLib for IMorpho;
+
+    /* IMMUTABLES */
+
+    IMorpho public immutable morpho;
+
+    /* CONSTRUCTOR */
+
+    /// @notice Constructs the contract.
+    /// @param morphoAddress The address of the Morpho Blue contract.
+    constructor(address morphoAddress) {
+        require(morphoAddress != address(0), "Invalid Morpho address");
+        morpho = IMorpho(morphoAddress);
+    }
+
+    /*  VIEW FUNCTIONS */
+
+    /// @notice Calculates the health factor of a user in a specific market.
+    /// @param marketParams The parameters of the market.
+    /// @param id The identifier of the market.
+    /// @param user The address of the user whose health factor is being calculated.
+    /// @return healthFactor The calculated health factor.
+    function userHealthFactor(MarketParams memory marketParams, Id id, address user)
+        public
+        view
+        returns (uint256 healthFactor)
+    {
+        uint256 collateralPrice = IOracle(marketParams.oracle).price();
+        uint256 collateral = morpho.collateral(id, user);
+        uint256 borrowed = morpho.expectedBorrowAssets(marketParams, user);
+
+        uint256 maxBorrow = collateral.mulDivDown(collateralPrice, ORACLE_PRICE_SCALE).wMulDown(marketParams.lltv);
+
+        if (borrowed == 0) return type(uint256).max;
+        healthFactor = maxBorrow.wDivDown(borrowed);
+    }
+
+    /// @notice Calculates the health factor of a user after a virtual repayment.
+    /// @param marketParams The parameters of the market.
+    /// @param id The identifier of the market.
+    /// @param user The address of the user whose health factor is being calculated.
+    /// @param repaymentAmount The amount of assets to be virtually repaid.
+    /// @return healthFactor The calculated health factor after the virtual repayment.
+    function userHealthFactorAfterVirtualRepayment(
+        MarketParams memory marketParams,
+        Id id,
+        address user,
+        uint256 repaymentAmount
+    ) public view returns (uint256 healthFactor) {
+        uint256 collateralPrice = IOracle(marketParams.oracle).price();
+        uint256 collateral = morpho.collateral(id, user);
+        uint256 borrowed = morpho.expectedBorrowAssets(marketParams, user);
+
+        // Revert if repaymentAmount exceeds the borrowed amount
+        require(repaymentAmount <= borrowed, "Repayment amount exceeds borrowed amount");
```

Comment:
We could simply let the snippet revert with underflow just like in Morpho Blue (idrc)